### ISSUE TITLE
Fragging cannot inflate MvxFragment

### ISSUE THREAD

**Issue by Cheesebaron** [MvvmCross/MvvmCross#994]
Reported on StackOverflow: http://stackoverflow.com/questions/30031976/whats-the-proper-way-to-implement-fragments-in-mvvmcross-3-5-1

Imagine you have an `Activity` that wants to inflate a `MvxFragment` like so:

```
[Register("fragmenttest.views.FirstViewFragment")]
public class FirstViewFragment : MvxFragment
{
	public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
	{
		var ignore = base.OnCreateView(inflater, container, savedInstanceState);

		return this.BindingInflate(Resource.Layout.FirstViewFragment, null);
	}
}
```

and the AXML it tries to inflate:

```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:local="http://schemas.android.com/apk/res-auto"
    android:orientation="vertical"
    android:layout_width="fill_parent"
    android:layout_height="fill_parent">
    <fragment
        class="fragmenttest.views.FirstViewFragment"
        android:id="@+id/sub_login_form"
        android:layout_width="fill_parent"
        android:layout_height="wrap_content" />
</LinearLayout>
```

This goes boom as `MvxFragment` is not recognized as a `Fragment` by Android. Full stack trace:

```
05-05 07:44:14.283 I/MonoDroid( 3392): UNHANDLED EXCEPTION:
05-05 07:44:14.287 I/MonoDroid( 3392): Android.Views.InflateException: Exception of type 'Android.Views.InflateException' was thrown.
05-05 07:44:14.287 I/MonoDroid( 3392): at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () <IL 0x00011, 0x00047>
05-05 07:44:14.287 I/MonoDroid( 3392): at Android.Runtime.JNIEnv.CallObjectMethod (intptr,intptr,Android.Runtime.JValue[]) [0x00064] in /Users/builder/data/lanes/monodroid-mlion-monodroid-4.21-series/38ac51a9/source/monodroid/src/Mono.Android/src/Runtime/JNIEnv.g.cs:195
05-05 07:44:14.287 I/MonoDroid( 3392): at Android.Views.LayoutInflater.Inflate (int,Android.Views.ViewGroup,bool) [0x00043] in /Users/builder/data/lanes/monodroid-mlion-monodroid-4.21-series/38ac51a9/source/monodroid/src/Mono.Android/platforms/android-19/src/generated/Android.Views.LayoutInflater.cs:609
05-05 07:44:14.287 I/MonoDroid( 3392): at Cirrious.MvvmCross.Binding.Droid.BindingContext.MvxAndroidBindingContext.CommonInflate (int,Android.Views.ViewGroup,Cirrious.MvvmCross.Binding.Droid.Binders.IMvxLayoutInfactorFactory,bool) <IL 0x00030, 0x00127>
05-05 07:44:14.287 I/MonoDroid( 3392): at Cirrious.MvvmCross.Binding.Droid.BindingContext.MvxAndroidBindingContext.BindingInflate (int,Android.Views.ViewGroup,bool) <IL 0x00015, 0x000a2>
05-05 07:44:14.287 I/MonoDroid( 3392): at Cirrious.MvvmCross.Binding.Droid.BindingContext.MvxAndroidBindingContext.BindingInflate (int,Android.Views.ViewGroup) <IL 0x00004, 0x00039>
05-05 07:44:14.287 I/MonoDroid( 3392): at Cirrious.MvvmCross.Binding.Droid.BindingContext.MvxBindingContextOwnerExtensions.BindingInflate (Cirrious.MvvmCross.Binding.BindingContext.IMvxBindingContextOwner,int,Android.Views.ViewGroup) <IL 0x0000f, 0x000f5>
05-05 07:44:14.287 I/MonoDroid( 3392): at Cirrious.MvvmCross.Droid.Fragging.MvxFragmentActivity.SetContentView (int) <IL 0x00003, 0x0002f>
05-05 07:44:14.287 I/MonoDroid( 3392): at FragmentTest.Views.FirstView.OnCreate (Android.OS.Bundle) [0x00009] in C:\Users\tcielecki\Downloads\frag2\Views\FirstView.cs:17
05-05 07:44:14.287 I/MonoDroid( 3392): at Android.App.Activity.n_OnCreate_Landroid_os_Bundle_ (intptr,intptr,intptr) [0x00011] in /Users/builder/data/lanes/monodroid-mlion-monodroid-4.21-series/38ac51a9/source/monodroid/src/Mono.Android/platforms/android-19/src/generated/Android.App.Activity.cs:2179
05-05 07:44:14.287 I/MonoDroid( 3392): at (wrapper dynamic-method) object.be4ee29a-daea-47ab-8d47-7cc01a1803e6 (intptr,intptr,intptr) <IL 0x00017, 0x0001f>
An unhandled exception occured.

05-05 07:44:14.287 I/MonoDroid( 3392):   --- End of managed exception stack trace ---
05-05 07:44:14.287 I/MonoDroid( 3392): android.view.InflateException: Binary XML file line #1: Error inflating class fragment
05-05 07:44:14.287 I/MonoDroid( 3392): 	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:713)
05-05 07:44:14.287 I/MonoDroid( 3392): 	at android.view.LayoutInflater.rInflate(LayoutInflater.java:755)
05-05 07:44:14.287 I/MonoDroid( 3392): 	at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
05-05 07:44:14.287 I/MonoDroid( 3392): 	at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
05-05 07:44:14.287 I/MonoDroid( 3392): 	at fragmenttest.views.FirstView.n_onCreate(Native Method)
05-05 07:44:14.291 I/MonoDroid( 3392): 	at fragmenttest.views.FirstView.onCreate(FirstView.java:28)
05-05 07:44:14.291 I/MonoDroid( 3392): 	at android.app.Activity.performCreate(Activity.java:5231)
05-05 07:44:14.291 I/MonoDroid( 3392): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1087)
05-05 07:44:14.291 I/MonoDroid( 3392): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2148)
05-05 07:44:14.291 I/MonoDroid( 3392): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2233)
05-05 07:44:14.291 I/MonoDroid( 3392): 	at android.app.ActivityThread.access$800(ActivityThread.java:135)
05-05 07:44:14.291 I/MonoDroid( 3392): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
05-05 07:44:14.291 I/MonoDroid( 3392): 	at android.os.Handler.dispatchMessage(Handler.java:102)
05-05 07:44:14.295 I/MonoDroid( 3392): 	at android.os.Looper.loop(Looper.java:136)
05-05 07:44:14.295 I/MonoDroid( 3392): 	at android.app.ActivityThread.main(ActivityThread.java:5001)
05-05 07:44:14.295 I/MonoDroid( 3392): 	at java.lang.reflect.Method.invokeNative(Native Method)
05-05 07:44:14.295 I/MonoDroid( 3392): 	at java.lang.reflect.Method.invoke(Method.java:515)
05-05 07:44:14.295 I/MonoDroid( 3392): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
05-05 07:44:14.295 I/MonoDroid( 3392): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
05-05 07:44:14.295 I/MonoDroid( 3392): 	at dalvik.system.NativeStart.main(Native Method)
05-05 07:44:14.295 I/MonoDroid( 3392): Caused by: android.app.Fragment$InstantiationException: Trying to instantiate a class fragmenttest.views.FirstViewFragment that is not a Fragment
05-05 07:44:14.295 I/MonoDroid( 3392): 	at android.app.Fragment.instantiate(Fragment.java:585)
05-05 07:44:14.295 I/MonoDroid( 3392): 	at android.app.Fragment.instantiate(Fragment.java:561)
05-05 07:44:14.299 I/MonoDroid( 3392): 	at android.app.Activity.onCreateView(Activity.java:4778)
05-05 07:44:14.299 I/MonoDroid( 3392): 	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:689)
05-05 07:44:14.299 I/MonoDroid( 3392): 	... 19 more
05-05 07:44:14.299 I/MonoDroid( 3392): Caused by: java.lang.ClassCastException
05-05 07:44:14.299 I/MonoDroid( 3392): 	... 23 more
05-05 07:44:16.351 E/mono    ( 3392): 
05-05 07:44:16.351 E/mono    ( 3392): Unhandled Exception:
05-05 07:44:16.351 E/mono    ( 3392): Android.Views.InflateException: Exception of type 'Android.Views.InflateException' was thrown.
05-05 07:44:16.351 E/mono    ( 3392): at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () <IL 0x00011, 0x00047>
05-05 07:44:16.351 E/mono    ( 3392): at Android.Runtime.JNIEnv.CallObjectMethod (intptr,intptr,Android.Runtime.JValue[]) [0x00064] in /Users/builder/data/lanes/monodroid-mlion-monodroid-4.21-series/38ac51a9/source/monodroid/src/Mono.Android/src/Runtime/JNIEnv.g.cs:195
05-05 07:44:16.351 E/mono    ( 3392): at Android.Views.LayoutInflater.Inflate (int,Android.Views.ViewGroup,bool) [0x00043] in /Users/builder/data/lanes/monodroid-mlion-monodroid-4.21-series/38ac51a9/source/monodroid/src/Mono.Android/platforms/android-19/src/generated/Android.Views.LayoutInflater.cs:609
```

The line that is interesting here:

    Caused by: android.app.Fragment$InstantiationException: Trying to instantiate a class fragmenttest.views.FirstViewFragment that is not a Fragment

    Caused by: java.lang.ClassCastException

Seems like there is some mismatch between classes used maybe? Btw this is using the Fragging 3.5.1 nuget (FullFragging works)

**Comment by slodge** [comment]
Argh

Is this support package version mismatches? Or some java class generation mis-alignment in the new xam md hashing world?

Is it maybe related to https://github.com/MvvmCross/MvvmCross/pull/988 ?

**Comment by Cheesebaron** [comment]
According to this line: https://github.com/kjeremy/MvvmCross/blob/5fe6272e22de6fa5543150689144ae6aa2c4a319/Cirrious/Cirrious.MvvmCross.Binding.Droid/Binders/MvxBindingLayoutInflatorFactory.cs#L61

Nothing is done by MvvmCross with Fragments in the code @kjeremy introduced. Not sure if just checking for `fragment` as name is enough there?

**Comment by kjeremy** [comment]
That line was there before I touched the file.  I believe it has always been the case that with mvvmcross at least you can't put fragments in the axml.

**Comment by kjeremy** [comment]
Honestly I think there's something funky going on with our layout inflation.  For instance the change I made doesn't allow the support library to inflate ```EditText``` as ```AppCompatEditText``` like it should.  Instead you need to do something like [this](https://gist.github.com/kjeremy/5b76cb2eb82ff6caa046) which is essentially what AppCompat is doing under the hood which tells me that AppCompat's factory never gets a chance to run.  I think the reason is because ```MvxBindingLayoutInflatorFactory``` gets called before other factories that may be attached and it inflates **all** views using ```Activator``` so that the other factories don't get a chance to run.

I tried returning null for the widgets in the gist so that AppCompat's factory could run but it didn't work.

**Comment by slodge** [comment]
Maybe I'm being dim....

- What's the problem if Mvx inflates those views? 
- If there is a problem, then is it just a case of finding an alternative way to create the view in https://github.com/kjeremy/MvvmCross/blob/5fe6272e22de6fa5543150689144ae6aa2c4a319/Cirrious/Cirrious.MvvmCross.Binding.Droid/Binders/MvxBindingLayoutInflatorFactory.cs#L64 - e.g. in passing through a different factory which knows about other inflators?

Really not sure what is a bug in Xam and Android "stable" here... feels like we haven't changed things... so not sure we can fix some of these breaks... :/

**Comment by roubachof** [comment]
I have the same experience as jeremy, without injecting the tint-aware views (gist), no tint is applied to api < lollipop. You can tell right away because the activity title doesn't receive the correct style (which is medium 20sp, http://www.google.com/design/spec/style/typography.html#typography-styles). 

Maybe we should create an issue for this so people can be aware of this. Because looking at #988, one user could think that everything should work as expected.

**Comment by kjeremy** [comment]
@slodge I think the issue is that AppCompat swaps out widgets that it knows about on inflation and our factory doesn't allow that to happen since it inflates everything.  So every time a new version of appcompat is released we all need to reimplement the behavior.  I wonder if anyone uses Calligraphy and MvvmCross together (@Cheesebaron ?) and if so do they run into the same issue?

When you set a factory it's supposed to merge them.  Maybe we need to return ```null``` if we aren't inflating an ```Mvx``` class or something like that or if that would break built-in widgets with bindings  Not sure what the solution is.  Maybe it's a combination of the above and moving https://github.com/kjeremy/MvvmCross/blob/5fe6272e22de6fa5543150689144ae6aa2c4a319/Cirrious/Cirrious.MvvmCross.Binding.Droid/Binders/MvxBindingLayoutInflatorFactory.cs#L65-L68 out of the inflater.

**Comment by slodge** [comment]
If you return null, then how are you going to data-bind it?

This line is needed after the view is created by whoever creates it -https://github.com/kjeremy/MvvmCross/blob/5fe6272e22de6fa5543150689144ae6aa2c4a319/Cirrious/Cirrious.MvvmCross.Binding.Droid/Binders/MvxBindingLayoutInflatorFactory.cs#L67

I'm guessing this isn't going to be a huge problem to overcome... but I'm a little lost (maybe cos I'm not in there fighting this problem at the moment). Definitely good to get good techie descriptions of actual problems if we can - and I'm not sure this thread is making much sense to me - I think maybe different people are talking about different things.

This thread started from a stackoverflow question which didn't really seem to be about AppCompat and said "It works fine when I build to a physical device" - so is this just an emulator issue? Or just specific Android versions? Or is the initial post in this thread a red-herring... and not what the rest of the conversation has evolved into....

Thinking.... Maybe I need a cake :)

**Comment by kjeremy** [comment]
Sorry, back to the original issue: I don't know why the support fragment manager has issues.  It looks like @Cheesebaron fixed it by using FullFragging.  Either there's an issue with using the Support libraries or maybe I messed up my implementation of LayoutCompat.

**Comment by dbeattie71** [comment]
I'm using SupportFragmentManager and haven't had any issues after using [Register]

**Comment by cyrilcathala** [comment]
Is the `[Register]` mandatory ? Is there any other workaround ? 
I'd like to migrate a project to Xam.Android 5.1 + Mvx 3.5.1 but it's really a pain in the *** to tag each and every fragment ...

Is it due to Mvx mixing with the inflater behind the scenes ? Could it be solved by the rework done by @kjeremy (#1022). Because if I understood correctly, the `[Register]` should only serve scenarios where you use a "class=" declaration within the xml.

**Comment by kjeremy** [comment]
Just for the record I tested both native and support fragments and they both work fine on the 4.0 branch.
